Explicit indexes

benbovy opened the pull request:
<!-- Feel free to remove check-list items aren't relevant to your change -->

- [x] Closes many issues:
  - [x] closes #1366 
  - [x] closes #1408 
  - [x] closes #2489 
  - [x] closes #3432 
  - [x] closes #4542 
  - [x] closes #4955
  - [x] closes #5202
  - [x] closes #5645 
  - [x] closes #5691 
  - [x] closes #5697
  - [x] closes #5700 
  - [x] closes #5727
  - [x] closes #5953
  - [x] closes #6183
  - [x] closes #6313
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- New functions/methods are listed in `api.rst` (new `Index` and `Indexes` API not public yet)

Follow-up on #5636 (work in progress), supersedes #2195.

This is likely to be going big, sorry in advance! It'll be safer to make a release before merging this PR.

Current progress:

- [x] create (default) indexes using the `Index` classes
  - [x] refactor default indexes created when 1st accessing `.xindexes` or `.indexes`
- [x] support for non-default indexes (no public API yet)
- [x] remove multi-index virtual coordinates (replace it by regular coordinates)
- [x] refactor internal (text / html) formatting functions
- [x] internal refactor of location-based selection (`.isel()`)
- [x] internal refactor of label-based selection (`.sel()`)
- [x] internal refactor of `.rename()`
  - Some changes in behavior (see comments below)
    - see #4108
    - see #4107
    - see #4417
- [x] internal refactor of `set_index` / `reset_index`
- [x] internal refactor of `stack` / `unstack`
    - Some changes in behavior (see comments below) 
- [x] internal refactor of `Dataset.to_stacked_array`
- [x] internal refactor of `swap_dims`
- [x] internal refactor of `expand_dims`
- [x] internal refactor of alignment
- [x] internal refactor of `reindex` and `reindex_like`
- [x] internal refactor of `interp` and `interp_like`
- [x] internal refactor of merge
- [x] internal refactor of concat
- [x] internal refactor of computation
- [x] internal refactor of copy
- [x] internal refactor of `update`, `assign`, `__setitem__`, `del`, `drop_vars`, etc.
    - updates must not corrupt multi-coordinate indexes 
- [x] internal refactor of `set_coords` and `reset_coords`
- internal refactor of `drop_sel` and `drop_isel` (maybe later)
- [x] internal refactor of `pad`
- [x] internal refactor of `shift`
- [x] internal refactor of `roll`

TODO:

- [x] Uniformize Index API with Xarray's API
    - [x] rename `Index.query()` -> `Index.sel()`?
    - [x] rename `PandasMultiIndex.from_product()` -> `PandasMultiIndex.stack()`? Add `Index.stack()` and `Index.unstack()`.
    - [x] remove `Index.union()` and `Index.intersection()`
- [x] Use `Index.create_variables()` internally
    - [x] remove `PandasIndex.from_pandas_index()` and `PandasMultiIndex.from_pandas_index()` (use constructor + `.create_variables()` instead)
- [x] Review where `.xindexes` is used and use private API instead (`._indexes`) if possible for speed
    - [x] requires that `_indexes` always returns a mapping
- [x] Use `from __future__ import annotations` in `indexes.py`
- [x] Re-activate default indexes invariant check (with opt-out for some tests)


In next PRs:

- custom `Index.__repr__` and `Index._repr_inline_`
- add an `Indexes` section in `DataArray` / `Dataset` reprs
- update public API (`set_index`, `reset_index`, `drop_indexes`, `Dataset` and `DataArray` constructors, etc.)
- allow multi-dimensional variables with `name` in `var.dims`

benbovy commented:
At this point, I'm wondering whether we should depreciate a couple of things now or later:

- Passing a multi-index directly as a coordinate in Dataset / DataArray constructors 

Currently, level coordinates are implicitly created from the multi-index. Suggested behavior: treat the index as an array-like (single coordinate) by default and encourage passing it more explicitly, e.g., something like

```python
midx = pd.MultiIndex.from_arrays([[...], [...]], names=("foo", "bar"))
coords, index = xr.PandasMultiIndex.from_pandas_index(midx, dim=“x”)
ds = xr.Dataset(coords=coords, indexes={("foo", "bar"): index})
```

- `IndexVariable.level_names` and `IndexVariable.get_level_variable`

- Maybe #5732 too?

Illviljan commented on xarray/core/dataarray.py:
```diff
@@ -462,23 +460,40 @@ def _replace_maybe_drop_dims(
+    def _overwrite_indexes(
+        self,
+        indexes: Mapping[Hashable, Index],
+        coords: Optional[Mapping[Hashable, Variable]] = None,
```
```suggestion
        coords: Optional[Mapping[Any, Variable]] = None,
```
A small note here but using hashable doesn't seem to be such a good idea. See #5690.
And if you do drive-by fixes to tests, add some typing to them so mypy can check them as well.

benbovy replied:
Ah I see, thanks! I didn't pay attention enough to #5690 (I'm still not super familiar with all Python annotation subtleties to be honest).

benbovy replied:
> if you do drive-by fixes to tests, add some typing to them so mypy can check them as well.

I suggest doing this in follow-up PRs (except for new tests) as this one is already getting big and messy.

Illviljan replied:
Sure, all I'm saying that if you're looking at or editing a test add at least a quick `-> None` so mypy checks it.

benbovy commented:
@max-sixty  I'm trying to merge the main branch here but I'm struggling with a lot of mypy errors like 

```
xarray/core/_typed_ops.pyi:179: error: Self argument missing for a non-static method (or an invalid type for self)  [misc]
```

Do you know how to fix this?

benbovy commented:
I did run `pre-commit clean` but the errors are still there on my local repository. It's weird that all pre-commit hooks pass on CI.

Illviljan commented on xarray/tests/test_indexing.py:
```diff
@@ -63,30 +64,59 @@ def test_group_indexers_by_index(self) -> None:
-                idx = idx.to_pandas_index()
-            assert_array_equal(pos.get("x"), expected_pos)
-            assert_array_equal(idx, expected_idx)
+    def test_remap_label_indexers(self):
+        def test_indexer(
+            data,
+            x,
+            expected_pos,
+            expected_idx=None,
+            expected_vars=None,
+            expected_drop=None,
+            expected_rename_dims=None,
+        ):
```
```suggestion
    def test_remap_label_indexers(self) -> None:
        def test_indexer(
            data,
            x,
            expected_pos,
            expected_idx=None,
            expected_vars=None,
            expected_drop=None,
            expected_rename_dims=None,
        ) -> None:
```
The typing got lost in the merge?

benbovy replied:
No it is a temporary dirty fix until I can figure out why I get a `too many arguments` mypy error. I checked the calls to `test_indexer` below and didn't see anything wrong. Maybe it has to do with unpacked arguments, something similar to https://github.com/python/mypy/issues/9007?

Illviljan replied:
I think you can do: 

```python
        a, b = PandasIndex.from_pandas_index(pd.Index([-1, -2]), "three")
        test_indexer(
            mdata,
            ("a", 1),
            [True, True, False, False, False, False, False, False],
            a,
            b,
            ["x", "one", "two"],
            {"x": "three"},
        )
```
Unnecessary workaround but that's how it is with mypy. 

Sometimes I wonder if it's better to add another type checker and IF both of them fail, then the CI is allowed to fail.

Illviljan replied:
The better solution is to type the output of `from_pandas_index`, for example like this:
```python
    @classmethod
    def from_pandas_index(cls, index: pd.Index, dim: Hashable) -> Tuple[Any, Any]:
```
I didn't figure what the output was so I used Any instead, but you have a better feel for that.

benbovy commented:
I'm currently adding more type annotations to the `Index` (sub)classes. I'm not 100% sure what to do for the `query` method, though. Without type annotations, I was thinking about those signatures:

```python
class Index:
    def query(self, labels, **kwargs):
        raise NotImplementedError()

class PandasIndex(Index):
    def query(self, labels, method=None, tolerance=None):
        ...

class CustomIndex(Index):
    def query(self, labels, option1="value", option2=5):
        ...
```

However, this is unsafe and mypy complains (https://github.com/python/mypy/issues/7424).

Alternative options that work with type annotations:

1. add `# type: ignore[override]` in `Index` subclasses
2. add `**kwargs` in subclasses too (those are ignored)
3. use a fixed number of arguments, i.e., `Index.query(labels, query_options: Optional[Dict[str, Any]] = None)`

Maybe other alternatives?

I don't like 1 and 2, it does not look very nice. 3 makes default query option values less clear, unless we define them elsewhere as a convention (e.g., as a class property)?

shoyer commented:
My suggestion:

Don't include keyword arguments inside the base class:
```python
class Index:
    def query(self, labels):
        raise NotImplementedError()
```

Still include custom keyword arguments on subclasses.

Use `# type: ignore` inside xarray, when we call `query()`

Illviljan commented on xarray/core/indexing.py:
```diff
@@ -18,7 +31,148 @@
+
+
+def group_indexers_by_index(
+    obj: Union["DataArray", "Dataset"],
```
```suggestion
    obj: T_Xarray,
```
You can import that from https://github.com/pydata/xarray/blob/7bfee3eaa8fd731494cf6b406d6abb4bec061001/xarray/core/types.py#L24

benbovy commented:
This PR introduces some minor changes in behavior (no API change), mostly related to somewhat tricky workarounds to the limitations of the “index/dimension” coordinates concept that we no longer need with explicit indexes.

I'll detail them below. A couple of those changes are bug fixes so we can safely make them now. For the other changes I'm not sure how best to proceed. Keeping the current behavior may in some cases require additional implementation effort that I'm not sure it's worth doing if no one relies on this weird behavior. Any thoughts @pydata/xarray?

---

1. `.rename_*`

1.1 Indexes are now preserved when renaming dimensions or coordinates

```python
ds = xr.Dataset({"x": ("x", [0, 1, 2])})
renamed = ds.rename_dims({"x": "x_new"})

# Before
"x" in renamed.indexes   # False
# Now
"x" in renamed.indexes   # True
```

---

2. `set_index`

2.1 Coordinate(s) dtype is preserved when setting new (multi-)indexes

```python
da = xr.DataArray(
    [0, 1, 3, 4],
    dims="x",
    coords={"x": [0, 1, 0, 1], "y": ("x", ["a", "b", "a", "b"])}
)
print(da)
# <xarray.DataArray (x: 4)>
# array([0, 1, 3, 4])
# Coordinates:
#   * x        (x) int64 0 1 0 1
#     y        (x) <U1 'a' 'b' 'a' 'b'
```
```python
indexed = da.set_index(xy=["x", "y"])

# Before
indexed.y.dtype    # dtype('O')
# Now
indexed.y.dtype    # dtype('<U1')
```
```python
indexed = da.set_index(x="y")

# Before
indexed.y.dtype    # dtype('O')
# Now
indexed.y.dtype    # dtype('<U1')
```

2.2 Setting a new single index for a new dimension raises an error

New dimension names allowed in `set_index` is useful to create new multi-indexes from scratch and avoid dimension name conflicts for its levels. However, new dimension names is also allowed for single indexes and resulted in this case to weird dimension renaming, now it raises an error.

```python
# Before (bug)
da.set_index(y="x")
# <xarray.DataArray (x: 4)>
# array([0, 1, 2, 3])
# Coordinates:
#   * y        (y) int64 0 1 0 1
# Dimensions without coordinates: x

# Now
da.set_index(y="x")
# ValueError: try setting an index for dimension 'y' 
# with variable 'x' that has dimensions ('x',)
```

---

3. `reset_index`

3.1 Coordinate (level) names are preserved when (partially) resetting a multi-index

```python
indexed = da.set_index(xy=["x", "y"])
indexed
# <xarray.DataArray (xy: 4)>
# array([0, 1, 2, 3])
# Coordinates:
#   * xy       (xy) object MultiIndex
#   * x        (xy) int64 0 1 0 1
#   * y        (xy) <U1 'a' 'a' 'b' 'b'
```

```python
# Before
indexed.reset_index("x")
# <xarray.DataArray (xy: 4)>
# array([0, 1, 2, 3])
# Coordinates:
#   * xy       (xy) object 'a' 'a' 'b' 'b'
#     x        (xy) int64 0 1 0 1

# Now
indexed.reset_index("x")
# <xarray.DataArray (xy: 4)>
# array([0, 1, 2, 3])
# Coordinates:
#     xy       (xy) object MultiIndex
#     x        (xy) int64 0 1 0 1
#   * y        (xy) <U1 'a' 'a' 'b' 'b'
```

(note: the `xy` coordinate in the latter result has `MultiIndex` but is actually not part anymore of any xarray index in the returned DataArray. We'll need to remove the `MultiIndex` inline repr in favor of the tuple (level values) coordinate, perhaps when we'll add an `Indexes` section to the xarray objects repr. Eventually we could also get rid of this tuple coordinate and keep the coordinates of the multi-index levels only)

3.2 Single index coordinates that are reset (but kept) are not renamed

```python
# Before
da.reset_index("x")
# <xarray.DataArray (x: 4)>
# array([0, 1, 2, 3])
# Coordinates:
#     y        (x) <U1 'a' 'a' 'b' 'b'
#     x_       (x) int64 0 1 0 1
# Dimensions without coordinates: x

# Now
da.reset_index("x")
# <xarray.DataArray (x: 4)>
# array([0, 1, 2, 3])
# Coordinates:
#     x        (x) int64 0 1 0 1
#     y        (x) <U1 'a' 'a' 'b' 'b'
```

3.3 Fix bug when trying to reset a non-dimension (and non-level) coordinate

```python
# Before
da.reset_index("y")
# <xarray.DataArray (x: 4)>
# array([0, 1, 2, 3])
# Coordinates:
#   * x        (x) int64 0 1 0 1
#     y_       (y) object 'a' 'a' 'b' 'b'

# Now
da.reset_index("y")
# ValueError: ('y',) are not coordinates with an index
```

benbovy commented:
Also from discussion with @shoyer:

It would make sense (in a follow-up PR) to depreciate the current `set_index` API (i.e., set new indexes per dimension) in favor of setting only one index with:

- one or more variable names
- optionally the index type (by default a pandas index or multi-index is created)
- optionally one or more index build options to pass to the custom index (pandas indexes have no build option).

This departs from `pandas.DataFrame.set_index` API but it's more adapted to xarray's new data model with explicit indexes.

Currently Xarray also allows creating (multi-)indexes from existing (multi-)indexes, either with `.set_index({...}, append=True)`, `.reset_index(["not", "all", "levels"])` and `.reorder_levels()`. This API is specific to pandas (multi-)indexes so maybe it would make sense to depreciate it? One advantage over simply drop the index and re-create it from scratch is a possible gain in performance, which probably makes sense in pandas but we're not sure if it's worth the extra complexity here. Unless this kind incremental index build would be useful for other, custom indexes too?

(we should probably discuss this further in a new issue)

benbovy commented:
With the new renaming behavior that doesn't add or drop any index, It seems pretty safe to me to allow renaming a dimension even when the new name corresponds to an existing variable (but not an existing dimension). FYI, this was forbidden in #3645. @keewis any thoughts on this?

benbovy commented:
In the last commits I refactored `stack` / `unstack`.

One change is that a multi-index is not *always* created with `stack`. It is created only if each of the dimensions to stack together have *one and only one* coordinate with a pandas index (this could be a non-dimension coordinate).

This could maybe address #5202, since we could simply drop the indexes before stacking the dimensions in order to avoid the creation of a multi-index. I don't think it's a big breaking change either unless there are users who rely on default multi-indexes with `range` (0, 1, 2...) levels. Looking at #5202, however, those default multi-indexes seem more problematic than something really useful, but I might be wrong here. Also, range-based indexes can still be created explicitly before stacking the dimensions if needed.

Another consequence is that `stack` is not *always* reversible, since `unstack` still requires a pandas multi-index (*one and only one* multi-index per dimension to unstack).

benbovy commented:
To properly test the changes in behavior detailed above, we need to also check for the indexes when comparing two xarray objects. I opened #5812 to discuss about that issue.

keewis commented:
> FYI, this was forbidden in #3645

The discussion for that is in #3438. In addition to renaming, `rename*` did allow to replace other variables, which I think we should only allow when explicitly requested (e.g. by requiring `replace=True`), regardless of what it does with indexes (I think @dcherian was trying to fix a bug related to that in #4108).

benbovy commented:
Yep I've seen those issues.

There's a lot of things going on in #3438. I wonder if the goal of chaining `rename_dims`, `set_index`, `stack`, etc. in the example shown in that issue wasn't ultimately to circumvent the limitation of dimension coordinates, which is not necessary anymore with explicit indexes. More particularly, neither `rename_dims` nor `stack` should cause drop-in replacement of other variables as a side effect in this PR.

Regarding #4108, I think this is not necessary anymore with explicit indexes. Since indexes are allowed for non-dimension coordinates too, we can (and it's perhaps more consistent) to preserve any existing index before/after renaming.

pep8speaks commented:
Hello @benbovy! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

































* In the file [`xarray/tests/test_indexing.py`](https://github.com/pydata/xarray/blob/6437e13c86f45be6f51e5037aeeb123ceca15453/xarray/tests/test_indexing.py):

> [Line 2:1](https://github.com/pydata/xarray/blob/6437e13c86f45be6f51e5037aeeb123ceca15453/xarray/tests/test_indexing.py#L2): [F401](https://duckduckgo.com/?q=pep8%20F401) 'typing.Dict' imported but unused
> [Line 2:1](https://github.com/pydata/xarray/blob/6437e13c86f45be6f51e5037aeeb123ceca15453/xarray/tests/test_indexing.py#L2): [F401](https://duckduckgo.com/?q=pep8%20F401) 'typing.cast' imported but unused

benbovy commented on xarray/core/indexes.py:
```diff
@@ -305,68 +394,165 @@ def __init__(self, array: Any, dim: Hashable, level_coords_dtype: Any = None):
     @classmethod
-    def from_pandas_index(cls, index: pd.MultiIndex, dim: Hashable):
-        level_meta = {}
+    def from_variables_maybe_expand(
```
Hmm I'm not sure if this case is really worth extending the `from_variables` interface. I might be wrong, though.

This `PandasMultiIndex.from_variables_maybe_expand` constructor is specific to the current `set_index` API (with the `append` option) that we might depreciate in favor a new, more generic API for setting one index at a time (https://github.com/pydata/xarray/pull/5692#issuecomment-919836550), so we might eventually just drop this constructor.

For `set_index`, I guess (re)building a new index from scratch will be fine in most cases. In other operations like data selection it sometimes makes sense to create a new index from an existing one. But I suspect that those operations will rely on other `xarray.Index` methods than `from_variables`. 

Do you have other potential use cases in mind where extending `from_variables` could be useful? (note, we probably need to pass both the existing index (if it exists) and its corresponding coordinate variables).

benbovy commented on xarray/core/dataset.py:
```diff
@@ -3748,11 +3634,90 @@ def set_index(
         )
-        return self._replace_vars_and_dims(variables, coord_names=coord_names)
 
     def reset_index(
```
I totally agree!

xref #4366

Illviljan commented on xarray/core/coordinates.py:
```diff
@@ -16,11 +16,11 @@
-from .utils import Frozen, ReprObject, either_dict_or_kwargs
-from .variable import Variable
+from .utils import Frozen, ReprObject
+from .variable import calculate_dimensions, Variable
```
```suggestion
from .variable import Variable, calculate_dimensions
```

benbovy commented:
> I really appreciate your efforts here! This is a giant refactoring PR. It's very important work but I know it can be hard to land.

Thanks @shoyer! On a positive note (starting a new week after a long week-end helps), lots of important tasks are done now, there has been good progress.

 > I wonder if there are things we could do to make this easier to land -- maybe we could defer some of these fixes to a later PR, if we wait on disabling implicit indexes?

We'll need to make all tests pass, but yes I'll try to see if we can provide some temporary (dirty) fixes that at least work under the current (legacy) data model (dimension coordinates with pandas indexes).

dcherian commented:
@benbovy Thank you! I think everyone appreciates your hard work here, and also how frustrating this kind of work is :)

rabernat commented:
Just chiming in to say 💪 ! We see the work you are putting in @benbovy. I'm so excited to be using this feature. Is there a way I can help?

benbovy commented:
Thanks @rabernat (and others), I really appreciate your encouragement!

I think I start to see the light at the end of the tunnel here :). I have now refactored all the main parts of Xarray's internals that are using indexes and I managed to bring the failing tests down to a few. Now I'm going to try clearing all CI failures before working on the [remaining TODO items](https://github.com/pydata/xarray/pull/5692#issue-966983801). Although this is still a draft PR, it is now mostly ready for review I think. I'll mark it as ready when all CI is green.

Some of the re-implementations might still be improved, like `align` and `concat`. It's hard to know how best to do so before experimenting with custom indexes, though. We can still further iterate on this later...

IMO the best way to help will be to heavily test this branch (or the main branch once it is merged) against some real-world cases, as this PR will certainly introduce regressions that are not caught by the tests. A few release candidates might help in gathering feedback from the broader Xarray community. We'll also need to keep an eye on the asv benchmarks to ensure that there's no major regression in performance (I haven't checked yet).

This PR also intentionally introduces some minor changes in behavior (see my previous comments). I think those changes are welcome as they fix things that we can rather consider as bugs or limitations. That said, it is important to carefully check that they won't break Xarray users' (good or bad) habits.

And of course, once it is ready I'm eager to help anyone experiment with this feature (custom indexes)! I have a few ideas on convenient ways to implement custom selection / alignment behavior while mostly reusing the current Xarray indexing capabilities based on pandas indexes.

benbovy commented:
@keewis FYI in the two last commits I've made some changes to fix failing tests in `test_units.py`, which are related to the implicit creation of indexes and coordinates from a multi-index given in `Dataset` or `DataArray` constructors.

That may not be a proper fix. The problem is quite tricky though, since multi-indexes may be passed - directly or wrapped in DataArrays (as data or via their coordinates) - either to `data_vars` or `coords` (it may even be passed to both at the same time). It gets even more complicated now that multi-indexes in DataArrays have multiple coordinates. That's a lot of implicit behavior that IMHO we should depreciate as soon as possible in favor of a more explicit way to pass (multi-)indexes to the `Dataset` and `DataArray` constructors.

keewis commented:
thanks, @benbovy, those changes look good to me for now. After this is merged I'll have to update the tests anyways.

In the meantime, if you find any tests where the behavior changed so much that it is tricky to fix them I'd be fine with xfailing / skipping them for now.

dcherian commented:
ALL GREEN! 😱

benbovy commented:
> ALL GREEN! 😱

Yeah it's good to see all those green checks :) 

> if you find any tests where the behavior changed so much that it is tricky to fix them I'd be fine with xfailing / skipping them for now.

Some broken tests in `test_units` actually helped finding issues that are not directly related those tests!

keewis commented on xarray/core/formatting.py:
```diff
@@ -428,13 +395,14 @@ def _mapping_repr(
+        summarizer=summarize_variable,
         expand_option_name="display_expand_coords",
         col_width=col_width,
+        indexes=coords.indexes,
```
This raises an error for custom indexes (which are not convertible to `pandas`). Shouldn't it be this instead:
```suggestion
        indexes=coords.xindexes,
```

keewis reviewed (commented):
I tried implementing a [custom index](https://github.com/keewis/xarray-custom-indexes), and it works, at least for `sel`! It was also a lot easier than I had expected. Great job, @benbovy!

Formatting aside, the most important issue I had was passing the custom index to `DataArray`, which at the moment requires the use of `fastpath=True` (or am I missing something?)

benbovy commented:
@keewis great! I actually haven't tried it yet, so it's good to know. I wonder how it goes when applying other operations for which the custom index has no implementation. It should either drop the index, propagate the index or raise an error depending on the cases. 

> Formatting aside, the most important issue I had was passing the custom index to DataArray, which at the moment requires the use of fastpath=True (or am I missing something?)

Yes, there's no public API yet for setting (custom) indexes explicitly. This can be done in a follow-up PR.

keewis commented:
> I wonder how it goes when applying other operations for which the custom index has no implementation.

let's try it:
- `arr.roll(x=5)`: index is not dropped, the result for `sel(indexers)` is different from before the `roll` (I guess that makes sense? `pandas` indexes behave the same way...)
- `copy`: raises `NotImplementedError`
- `arr.stack(z=("x", "y"))`: index is dropped

I don't have ideas on how to test the other operations right now.

benbovy commented:
In [df31fa8](https://github.com/pydata/xarray/pull/5692/commits/df31fa81164af55c7d4fd09a285b75bb49f441eb) I have re-enabled the default indexes invariant check, which is now optional (enabled by default).

The changes in behavior in this PR (mostly related to `rename` and `reset_index`) can thus be tracked looking at the `check_default_indexes=False` option set for `assert_*` functions in the tests.

benbovy commented:
> let's try it:

>    arr.roll(x=5): index is not dropped, the result for sel(indexers) is different from before the roll (I guess that makes sense? pandas indexes behave the same way...)
    copy: raises NotImplementedError
    arr.stack(z=("x", "y")): index is dropped
    arr.isel(x=slice(5, 10), y=slice(10, 20)): index is dropped

For `arr.roll`, the index is dropped if `roll_coords=True` and `CustomIndex.roll()` is not implemented. Otherwise, the index is kept as-is since the coordinates are not affected.

For `copy`, we should probably use `copy.copy()` by default instead of raising an error. 

> I don't have ideas on how to test the other operations right now.

Yeah this is something that we can do progressively as we're experimenting with custom indexes. But this first try is already great.

keewis commented:
> For `arr.roll`, the index is dropped if `roll_coords=True` and `CustomIndex.roll()` is not implemented.

right, I can confirm that, too

Illviljan commented:
> I’m not sure what’s happening with the failing benchmark test `unstacking.UnstackingSparse.time_unstack_to_sparse_2d`, it seems flaky (or I missed something when trying to profile this benchmark + setup).

The benchmark triggers an error if it is at least 1.5x slower than main. If I remember correctly I believe `unstacking.UnstackingSparse.time_unstack_to_sparse_2d` has been around 1.45 in tests that have passed. So that one might be a little slower now but I wouldn't worry about that too much.

benbovy commented:
> The benchmark triggers an error if it is at least 1.5x slower than main. If I remember correctly I believe unstacking.UnstackingSparse.time_unstack_to_sparse_2d has been around 1.45 in tests that have passed. So that one might be a little slower now but I wouldn't worry about that too much.

Ah now I see that it is likely because this branch makes a copy of the pandas index (+ a second one that is not necessary and that I'll fix). Although making a shallow copy of a pandas index is rather cheap, it significantly affects the benchmark results here because of the small size of the index. For larger index sizes it becomes irrelevant:

This branch:

- size = 100

```python
>>> %timeit da_eye_2d.unstack(sparse=True)
680 µs ± 3.34 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

- size = 100 000

```python
>>> %timeit da_eye_2d.unstack(sparse=True)
49.2 ms ± 457 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Xarray stable 0.21.1 (with same versions of numpy, pandas and sparse):

- size = 100

```python
>>> %timeit da_eye_2d.unstack(sparse=True)
424 µs ± 13.1 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

- size = 100 000

```python
>>> %timeit da_eye_2d.unstack(sparse=True)
49.4 ms ± 797 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Illviljan commented on xarray/tests/test_indexes.py:
```diff
@@ -46,94 +144,200 @@ def test_from_variables(self) -> None:
-            index.query({"x": {"one": 0}})
+            index.sel({"x": {"one": 0}})
+
+    def test_query_boolean(self) -> None:
```
```suggestion
    def test_sel_boolean(self) -> None:
```

Illviljan commented on xarray/tests/test_indexes.py:
```diff
@@ -46,94 +144,200 @@ def test_from_variables(self) -> None:
+            actual.dim_indexers["x"], expected_dim_indexers["x"]
+        )
 
     def test_query_datetime(self) -> None:
```
```suggestion
    def test_sel_datetime(self) -> None:
```

Illviljan commented on xarray/tests/test_indexes.py:
```diff
@@ -46,94 +144,200 @@ def test_from_variables(self) -> None:
+        actual = index.sel({"x": index.to_pandas_index().to_numpy()[1]})
+        assert actual.dim_indexers == expected_dim_indexers
 
     def test_query_unsorted_datetime_index_raises(self) -> None:
```
```suggestion
    def test_sel_unsorted_datetime_index_raises(self) -> None:
```

Illviljan commented on xarray/tests/test_dataarray.py:
```diff
@@ -2156,35 +2184,46 @@ def test_dataset_math(self):
-
-        actual = orig.stack(z=["x", "y"]).unstack("z").drop_vars(["x", "y"])
+        # no default range index anymore
+        # a = orig[:0, :1].stack(dim=("x", "y")).dim.to_index()
+        # b = pd.MultiIndex(
+        #     levels=[pd.Index([], np.int64), pd.Index([0], np.int64)],
+        #     codes=[[], []],
+        #     names=["x", "y"],
+        # )
+        # pd.testing.assert_index_equal(a, b)
```
Uncomment or remove?

Illviljan commented on xarray/core/indexes.py:
```diff
@@ -387,103 +827,392 @@ def query(self, labels, method=None, tolerance=None):
+        return len(self._id_coord_names[self._coord_name_id[key]]) > 1
+
+    def get_all_coords(
+        self, key: Hashable, errors: str = "raise"
```
```suggestion
        self, key: Hashable, errors: Literal["raise", "ignore"] = "raise"
```
Can use Literal here and then I think we have done enough and can remove the runtime checks.

Illviljan reviewed (commented):
Mostly nits and some reminders as I can't say I understand everything here. 

I'm curious how a version with dask arrays would look like.

benbovy commented:
Thanks for the review @Illviljan! I addressed most of your comments.

Regarding the use of Literal, `join` options and `{"raise", "ignore"}` option values are used in several places, directly or indirectly related to this PR. I would suggest to take care of this in a follow-up PR.

benbovy commented:
Something I was worried about was the number of open PRs that would need to be updated after this one gets merged eventually. I quickly went through the most recent PRs and it seems to be ok, although there may be some PRs that I overlooked.

shoyer commented on xarray/core/concat.py:
```diff
@@ -506,22 +512,52 @@ def ensure_common_dims(vars):
-        elif k in result_vars:
+                raise ValueError(f"{name!r} is not present in all datasets.")
+
+            # Try concatenate the indexes first, silently fallback to concatenate
+            # the variables when no index is found on all datasets or when the
+            # 1st index doesn't implement concat.
```
Could we eliminate the fallback case here?

I guess `concat` is internally pretty robust at present, but requiring matching indexes would keep things simpler.

benbovy replied:
I agree that it would keep things simpler, at least for now.

More generally, I'm wondering if requiring matching indexes for alignment, merge and concat wouldn't be too restrictive / not practical.

For example, let's say that we want to concatenate a sequence of Xarray objects along a dimension `x` where some of them have a geospatial raster (meta-)index, e.g., for dimensions `x` and `y`, and all the others have decoupled pandas indexes for `x` and `y`. Which behavior do we want?

- A. Raise an error (no matching index along `x`, which is actually raised when aligning the objects before concatenation)
- B. Concatenate the coordinate variables along `x`, merge along the other dimensions and return the concatenated Dataset or DataArray with default pandas indexes

Option B would perhaps be more convenient, but significantly harder to implement! Although it is also easy to write `xr.concat([ds.reset_index("x") for ds in datasets], "x")`.

Back to your comment, I guess that you are referring to the 1st fallback and not the one when `index.concat` is not implemented?

shoyer replied:
I think it would probably be OK to remove both fall-backs here:

1. Require that all indexes exist if at least one exists.
2. Require implementing `Index.concat`.

Right now you are simply dropping indexes in this case, which seems non-ideal.

benbovy replied:
Removing both fall-backs breaks this test:

```python
def test_concat_promote_shape(self) -> None:
    # mixed dims within variables
    objs = [Dataset({}, {"x": 0}), Dataset({"x": [1]})]
    actual = concat(objs, "x")
```

What should we do?

- A. Implicitly create a ``PandasIndex`` for any scalar coordinate to concatenate, then concatenate the indexes with `Index.concat`
- B. Fallback to `Variable.concat` (it will create a new, default `PandasIndex` from the concatenated variable, when creating the resulting Dataset object)

Option A will likely not work if the other existing indexes are not all `PandasIndex` instances. Option B is a bit more flexible, although we should be careful when to fall-back as I agree dropping indexes is non-ideal, e.g., the example below:

```python
objs = [xr.Dataset(coords={"z": ("y", [2, 3])}), xr.Dataset(coords={"z": ("x", [4, 5])})]
# Let's set an index for the 1st object only (no public API yet)
objs[0]._indexes["z"] = PandasIndex([2, 3], "y")

# this should probably raise an error
# instead of concatenate "z", promote its dims to ("y", "x"), and drop the index
actual = xr.concat(objs, "x")
```

shoyer replied:
OK, I'm pretty sure I wrote that test! And now I'm doubting whether that behavior is really desirable, though I'm sure it may be more convenient than getting an error.

I would consider limiting the behavior to (A) for now (only supporting PandasIndex), and maybe deprecating it later?

shoyer reviewed (commented):
a few small comments

benbovy commented on xarray/core/alignment.py:
```diff
@@ -1,88 +1,557 @@
+            )
+        )
+
+    def align(self) -> tuple[DataAlignable, ...]:
+        if not self.indexes and len(self.objects) == 1:
+            # fast path for the trivial case
+            (obj,) = self.objects
+            return (obj.copy(deep=self.copy),)
+
+        self.find_matching_indexes()
+        self.find_matching_unindexed_dims()
+        self.assert_no_index_conflict()
+        self.align_indexes()
+        self.assert_unindexed_dim_sizes_equal()
+
+        if self.join == "override":
+            return self.override_indexes()
+        else:
+            return self.reindex_all()
```
Yes, that would be better indeed.

The whole point of the `Aligner` class is to re-organize the complex logic of alignment into more readable steps and in a more composable way, i.e., somewhat like the Builder pattern although here I don't think we'll ever need to call any of the intermediate methods (`find_matching_indexes`, etc.) externally.

Alternatively, with could make this point even clearer by renaming `Aligner` to `Align` and `Aligner.align` to `Align.__call__`. Not sure I prefer this option, though. `Align(...)()` doesn't look very nice and there's already an `align` function.

shoyer commented:
I would like to merge this PR very soon so it can get testing before the next release. If anyone has any remaining concerns, please speak up!

keewis commented on xarray/core/variable.py:
```diff
@@ -3033,3 +3020,43 @@ def assert_unique_multiindex_level_names(variables):
+    in new variables.
+
+    """
+    for name, var in new_variables.items():
+        old_var = old_variables.get(name)
+        if old_var is not None:
+            var.attrs = {**old_var.attrs, **var.attrs}
+            var.encoding = {**old_var.encoding, **var.encoding}
```
I wonder if we should do a shallow copy of each new variable instead of modifying them directly, and return that? I usually try to avoid functions with side-effects as much as possible...

benbovy replied:
I removed that function, which was only used within `Dataset._overwrite_indexes`. I still update the input `variables` in the latter instead of making a shallow copy because those variables won't be reused elsewhere and also because copying indexed coordinate variables do not currently preserve the identity of the wrapped pandas indexes (in this refactor we rely on this identify to find the indexes and their corresponding coordinates). This would need a deeper refactor of `IndexVariable.copy()`...

keewis commented on xarray/core/dataset.py:
```diff
@@ -2358,24 +2210,22 @@ def isel(
-            variables[var_name] = var_value
-            dims.update(zip(var_value.dims, var_value.shape))
+
+        indexes, index_variables = isel_indexes(self.xindexes, indexers)
```
```suggestion
        indexes, index_variables = isel_indexes(self._indexes, indexers)
```

keewis commented on xarray/core/dataset.py:
```diff
@@ -2394,29 +2244,22 @@ def _isel_fancy(
 
         variables: dict[Hashable, Variable] = {}
-        indexes: dict[Hashable, Index] = {}
+        indexes, index_variables = isel_indexes(self.xindexes, valid_indexers)
```
```suggestion
        indexes, index_variables = isel_indexes(self._indexes, valid_indexers)
```

keewis commented on xarray/core/dataset.py:
```diff
@@ -3294,7 +3194,14 @@ def interp_like(
+
+        # pick only dimension coordinates with a single index
+        coords = {}
+        other_indexes = other.xindexes
```
```suggestion
        other_indexes = other._indexes
```

keewis commented on xarray/core/dataset.py:
```diff
@@ -3335,28 +3242,34 @@ def _rename_vars(self, name_dict, dims_dict):
-        return indexes
+        variables = {}
+
+        for index, coord_names in self.xindexes.group_by_index():
```
```suggestion
        for index, coord_names in self._indexes.group_by_index():
```

keewis commented on xarray/core/dataset.py:
```diff
@@ -3780,11 +3693,87 @@ def set_index(
+                    + " variable(s) do not exist"
+                )
+
+            current_coord_names = self.xindexes.get_all_coords(dim, errors="ignore")
```
```suggestion
            current_coord_names = self._indexes.get_all_coords(dim, errors="ignore")
```

keewis commented on xarray/core/dataset.py:
```diff
@@ -3845,61 +3876,149 @@ def reorder_levels(
+                and (
+                    # stack: must be a single coordinate index
+                    not multi
+                    and not self.xindexes.is_multi(name)
```
```suggestion
                    and not self._indexes.is_multi(name)
```

keewis commented on xarray/core/dataset.py:
```diff
@@ -3845,61 +3876,149 @@ def reorder_levels(
+
+        # drop indexes of stacked coordinates (if any)
+        for name in stacked_var_names:
+            drop_indexes += list(self.xindexes.get_all_coords(name, errors="ignore"))
```
```suggestion
            drop_indexes += list(self._indexes.get_all_coords(name, errors="ignore"))
```

keewis commented on xarray/core/dataset.py:
```diff
@@ -6030,29 +6180,27 @@ def roll(
+        unrolled_vars: tuple[Hashable, ...]
+
+        if roll_coords:
+            indexes, index_vars = roll_indexes(self.xindexes, shifts)
```
```suggestion
            indexes, index_vars = roll_indexes(self._indexes, shifts)
```

keewis reviewed (commented):
I've got a few minor comments for you, but otherwise this looks good to me.

Feel free to ignore the `xindexes` -> `_indexes` suggestions if they don't make sense

benbovy commented on xarray/core/concat.py:
```diff
@@ -506,22 +512,52 @@ def ensure_common_dims(vars):
+                    )
+                    result_vars[name] = combined_var
+                else:
+                    idx_vars = datasets[0].xindexes.get_all_coords(name)
```
`_indexes` returns a plain dictionary `dict[Hashable, Index]`, while `xindexes` returns an instance of `Indexes`. The latter provides a mapping interface as well as some convenient methods to facilitate access to the indexes and their related coordinate variables.

I used `_indexes` when possible (for performance), but here and in your comments below I used `xindexes` as we need the `Indexes` API.

keewis replied:
okay, thanks

benbovy commented on xarray/core/dataarray.py:
```diff
@@ -2099,6 +2136,14 @@ def stack(
+            If False, don't create any index.
+            If None, create a multi-index only if exactly one single (1-d) coordinate
+            index is found for every dimension to stack.
+        index_cls: class, optional
```
Could we address this later, e.g., when making indexes (i.e., `Index`, `PandasIndex` and `PandasMultiIndex`) part of the public API? We can also remove it until then (although it would prevent hacking with custom indexes).

keewis replied:
of course! This is not public API yet, so this will not be rendered for now. We just have to make sure we don't forget.

benbovy replied:
> We just have to make sure we don't forget.

I updated #6293.

benbovy commented:
Ok, thanks @shoyer and @keewis for your comments!

The two failing CI checks are unrelated (see https://github.com/spatialaudio/nbsphinx/issues/639 for the doc build).

keewis commented:
the doc build actually fails because `nbconvert` relied on the implicit dependency of `nbformat` on `ipython_genutils`, but that got removed in the most recent release of `nbformat`. This is fixed in #6350.

shoyer commented:
OK, I think we’re good to go here?

keewis commented:
sounds good to me!

dcherian commented:
Yes!

shoyer commented:
OK, in it goes! Big thanks to @benbovy for seeing this through :)

jhamman commented:
Huge! Amazing effort here @benbovy!

max-sixty commented:
Huge success! Thank you @benbovy !
